keps/sig-instrumentation/draft-20190404-kubernetes-control-plane-metrics-stability.md

```diff
@@ -0,0 +1,294 @@
+
+When a stable metric undergoes the deprecation process, we are signaling that the metric will eventually be deleted. The lifecyle looks roughly like this (each stage represents a kubernetes release):
+
+__Stable metric__ -> __Deprecated metric__ -> __Hidden metric__ -> __Deletion__
+
+__Deprecated__ metrics have the same stability guarantees of their counterparts. If a stable metric is deprecated, then a deprecated stable metric is guaranteed to *not change*. When deprecating a stable metric, a future kubernetes release is specified as the point from which the metric will be considered deprecated.
+
+```go
+var someCounter = kubemetrics.CounterOpts{
+    Name: "some_counter",
+    Help: "this counts things",
+    StabilityLevel: kubemetrics.STABLE,
+    DeprecatedVersion: "1.15", // this metric is deprecated when the kubernetes version == 1.15
+}
+````
+
+__Deprecated__ metrics will have their description text prefixed with a deprecation notice string '(Deprecated from x.y)' and a warning log will be emitted during metric registration (in the spirit of the official [kubernetes deprecation policy](https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-flag-or-cli)).
+
+Before deprecation:
+
+```text
+# HELP some_counter this counts things
+# TYPE some_counter counter
+some_counter 0
+```
+
+During deprecation:
+```text
+# HELP some_counter (Deprecated from 1.15) this counts things
+# TYPE some_counter counter
+some_counter 0
+```
+Like their stable metric counterparts, deprecated metrics will be automatically registered to the metrics endpoint.
+
+On a subsequent release (when the metric's deprecatedVersion is equal to current_kubernetes_version - 1)), a deprecated metric will become a __hidden metric__. _Unlike_ their deprecated counterparts, hidden metrics will __*no longer be automatically registered*__ to the metrics endpoint (hence hidden). However, they can be explicitly enabled through a command line flag on the binary (i.e. '--enable-hidden-metrics=really_deprecated_metric'). This is to provide cluster admins an escape hatch to properly migrate off of a deprecated metric, if they were not able to react to the earlier deprecation warnings. Hidden metrics should be deleted after one release.
+
+## Design Details
+
+### Test Plan
```

Comment:
Do you think we can come up with a reasonable smoke test that validates that non-deprecated stable metrics aren't accidentally removed?

Reply:
Yes. One possibility is that we record the stable metrics on release bumps. We can then run static analysis against the code base to determine all current stable metric names to compare against our stored version and verify no rule violations.

Reply:
That works for me.